### PR TITLE
`DataflowTemplatedJobStartOperator` fix overwriting of location with default value, when a region is provided.

### PR DESCRIPTION
**Pull request by VVildVVolf**
Fix overwriting of location with default value, when a region is provided.

### **TL/DR**
When the `DataflowTemplateOperator` is being migrated from Airflow 1 to Airflow 2, if `region` property is setup with no `location` passing - Fallback anyway throws a message:

_**The mutually exclusive parameter `location` and `region` key in `variables` parameter are both present. Please remove one.**_

New unit tests of operator might explain better - they are not going to pass without the fix.

### Details
The `DataflowTemplatedJobStartOperator` has a parameter `location` with non empty default value. This operator is using the `DataflowHook` object. `DataflowHook` makes a check: if `location` and `region` are setup together at the same time, than throws the exception (like above). Since `DataflowTemplatedJobStartOperator`'s `location` has a default value, it will always pass non-zero value to the `DataflowHook`.

At the same time `DataflowHook` has the same default value for its  `location` parameter, so `DataflowTemplatedJobStartOperator` can safely pass None value (Special unit test is also added, to make sure `DataflowHook` will use default value when `location` and `region` are not provided at the same time).

### Notices

The same fallback `_fallback_to_location_from_variables` is applied to 4 methods (at the moment of writing):

1. start_template_dataflow (fixed case)
2. start_java_dataflow (found only definition and unit test references)
3. start_python_dataflow (found only definition and unit test references)
4. is_job_dataflow_running (referenced by `BeamRunJavaPipelineOperator`, `DataflowTemplatedJobStartOperator` and `DataflowCreateJavaJobOperator` with explicitly no passing `location` and `region` at the same time)

So, since other places are not affected, the change can be pretty small and focused on `DataflowTemplatedJobStartOperator` only.

P.S. Looks like some other people had the same problem, for example:  https://stackoverflow.com/questions/70254851/dataflowtemplateoperator-job-failing-on-cloud-composer-after-upgrading-to-airflo

---

### CODE REVIEW

**Review by potiuk — APPROVED** [review]
LGTM

File: airflow/providers/google/cloud/operators/dataflow.py [hunk 1]
```diff
@@ -600,7 +600,7 @@ def __init__(
         options: dict[str, Any] | None = None,
         dataflow_default_options: dict[str, Any] | None = None,
         parameters: dict[str, str] | None = None,
-        location: str = DEFAULT_DATAFLOW_LOCATION,
+        location: str | None = None,
```
Comment (eladkal) [inline airflow/providers/google/cloud/operators/dataflow.py:603]: Is this a breaking change or a bug fix?

Comment (potiuk) [reply airflow/providers/google/cloud/operators/dataflow.py:603]: It's not breaking as the default location is assigned in hook when it is passed as null and there is no region defined (as far as I understand)

File: tests/providers/google/cloud/operators/test_dataflow.py [hunk 2]
```diff
@@ -534,6 +535,42 @@ def test_validation_deferrable_params_raises_error(self):
...
+            "poll_sleep": POLL_SLEEP,
+            "wait_until_finished": True,
+            "cancel_timeout": CANCEL_TIMEOUT,
+        }
+        operator = DataflowTemplatedJobStartOperator(**init_kwargs)
+        operator.execute(None)
+        assert dataflow_mock.called
+        _, kwargs = dataflow_mock.call_args_list[0]
+        assert kwargs["variables"]["region"] == TEST_REGION
+        assert kwargs["location"] == None
```
Comment (pankajastro) [inline tests/providers/google/cloud/operators/test_dataflow.py]:
```suggestion
        assert kwargs["location"] is None
```

**Review by pankajastro — APPROVED** [review]

File: airflow/providers/google/cloud/operators/dataflow.py (hunk 1 above)
Comment (hussein-awala) [reply airflow/providers/google/cloud/operators/dataflow.py:603]: I have the same concern, I think this needs to be tested

Comment (VVildVVolf) [reply airflow/providers/google/cloud/operators/dataflow.py:603]: It is bug fix. The additional tests are provided - the case is valid but will be broken without the fix. The default value is not necessary on the operator level, since it is being passed to the hook anyway (and it has default).

Comment (VVildVVolf) [reply airflow/providers/google/cloud/operators/dataflow.py:603]: Also additional level of protection can be provided at [_fallback_variable_parameter method](https://github.com/apache/airflow/blob/793d36f013efa191e7732faf6e624cdd1d83ebae/airflow/providers/google/cloud/hooks/dataflow.py#LL85C5-L85C33), but looks redundant since the default value is caught from the [DataflowHook.start_template_dataflow](https://github.com/apache/airflow/blob/793d36f013efa191e7732faf6e624cdd1d83ebae/airflow/providers/google/cloud/hooks/dataflow.py#L621) definition itself. Please let me know if it is useful - I can publish the PR

Comment (eladkal) [reply airflow/providers/google/cloud/operators/dataflow.py:603]: My question was for clarification of this change in the release notes.
I have no objection either way
